src/storage/src/render/upsert/types.rs

```diff
@@ -448,8 +448,6 @@ pub struct InMemoryHashMap {
 
 impl InMemoryHashMap {
     fn drain(&mut self) -> Drain<'_, UpsertKey, StateValue> {
-        // clearing the total_size when draining
-        self.total_size = 0;
```

Comment:
I'm not following why this isn't correct? The `self.state.drain()` will clear the hashmap and we want this value to match the resulting size of the hashmap, right?

If the caller of `drain()` needs access to the size of the hashmap before they call `drain`, maybe we should return it too `(iterator_size, self.state.drain())`?

Reply:
E.g. here https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/render/upsert/types.rs#L622 we have access to `in_memory_size` and if that's guaranteed to not change then we could use `put_stats.size_diff -= in_memory_size;`?

Reply:
updated!